src/lib/storage/table.cpp

```diff
@@ -125,11 +126,13 @@ uint32_t Table::max_chunk_size() const { return _max_chunk_size; }
 
 std::shared_ptr<Chunk> Table::get_chunk(ChunkID chunk_id) {
   DebugAssert(chunk_id < _chunks.size(), "ChunkID " + std::to_string(chunk_id) + " out of range");
+  std::unique_lock lock(_chunk_mutex);
```

Comment:
Our logic guarantees that we do not try to get the pointer of a chunk that is about to be physically deleted, correct? TSan cannot know about this, so I would rather suppress that warning (with a comment here) instead of locking the chunks. For fun, you could run the TPC-H in multi-threaded mode on a beefy machine and you will see that this largely kills the performance.

Reply:
I have recently watched [Herb Sutter's atomic weapons talks](https://www.youtube.com/watch?v=A8eCGOqgvH4) and I feel more insecure about my MT knowledge than ever. My takeaway is that there is no such thing as "our logic" because sequential consistency is limited to one thread - it only applies to multi-threading if you synchronize properly. Otherwise, here be dragons, or something.

I won't give my blessing to anything that doesn't synchronize via atomic or mutex and to nothing that relies on the impartiality of 64 bit reads, simply because I don't know enough about the topic. 

It seems unwise to me to silence TSAN on such a trivial matter, I mean this is just multiple threads rw-accessing the same vector. If TSAN detects a race here, there probably is one.

Reply:
The internet is reeeeeaaaaally sloooooow here, so I can't properly work with SSH. Maybe you could start the TPC-H MT benchmark for ac06e5d^ and ac06e5d and paste it here?

Reply:
Well it doesn't properly kill it, but I still think this hit is avoidable:

```
./hyriseBenchmarkTPCH --scheduler --clients 100 -o after.json -q 1,2,3,4,5,6,7,8,9,10

+-----------+---------------+-------+---------------+-------+--------+---------------------------------+
| Benchmark | prev. iter/s  | runs  | new iter/s    | runs  | change | p-value (significant if <0.001) |
+-----------+---------------+-------+---------------+-------+--------+---------------------------------+
| TPC-H 1   | 16.0326118469 | 962   | 15.3656196594 | 922   | -4%    |     (run time too short) 0.6157 |
| TPC-H 2   | 11.38326931   | 683   | 9.93315410614 | 596   | -13%   |     (run time too short) 0.1662 |
| TPC-H 3   | 181.402236938 | 10000 | 174.466079712 | 10000 | -4%    |     (run time too short) 0.0083 |
| TPC-H 4   | 97.333114624  | 5840  | 94.1878051758 | 5652  | -3%    |     (run time too short) 0.0000 |
| TPC-H 5   | 144.038482666 | 8643  | 143.321044922 | 8600  | -0%    |     (run time too short) 0.0876 |
| TPC-H 6   | 467.382324219 | 10001 | 410.694335938 | 10003 | -12%   |     (run time too short) 0.0000 |
| TPC-H 7   | 49.525642395  | 2972  | 48.1816558838 | 2891  | -3%    |     (run time too short) 0.3558 |
| TPC-H 8   | 16.5375728607 | 993   | 16.4296684265 | 986   | -1%    |                          0.6665 |
| TPC-H 9   | 68.806427002  | 4129  | 69.1422424316 | 4149  | +0%    |     (run time too short) 0.5791 |
| TPC-H 10  | 104.351455688 | 6262  | 103.053039551 | 6184  | -1%    |     (run time too short) 0.6114 |
| average   |               |       |               |       | -4%    |                                 |
+-----------+---------------+-------+---------------+-------+--------+---------------------------------+
```

I'll give this some thought in the next days.

Reply:
Not getting through with this as fast as I hoped for. Let's bring the version with locks in and look at this separately.